python/new_examples/constraint_callback.py

```diff
@@ -0,0 +1,46 @@
+"""An example of overriding the ConstraintSolver's default callback function."""
+import nimblephysics as nimble
+import torch
+
+
+def dummy_callback(world: nimble.simulation.World):
+    pass
+
+def lcp_callback(world: nimble.simulation.World):
+    world.getConstraintSolver().lcpSolveCallback(world)
+
+def frictionless_lcp_callback(world: nimble.simulation.World):
+    # Backup and remove friction.
+    friction_coefs = []
+    bodies = []
+    for i in range(world.getNumBodyNodes()):
+        body = world.getBodyNodeIndex(i)
+        bodies.append(body)
+        friction_coefs.append(body.getFrictionCoeff())
+        body.setFrictionCoeff(0.0)
+
+    # Frictionless LCP
+    lcp_callback(world)
+
+    # Restore friction.
+    for friction_coef, body in zip(friction_coefs, bodies):
+        body.setFrictionCoeff(friction_coef)
+
+def main():
+    world = nimble.loadWorld("../../data/skel/test/colliding_cube.skel")
+    state = torch.tensor(world.getState())
+    action = torch.zeros((world.getNumDofs()))
+    solver = world.getConstraintSolver()
+    callbacks = [None, dummy_callback, solver.lcpSolveCallback, lcp_callback, frictionless_lcp_callback]
+    for callback in callbacks:
+        if callback is not None:
+            solver.replaceSolveCallback(callback)
+            print(callback.__name__)
+        else:
+            print("None")
+        new_state = nimble.timestep(world, state, action)
```

Comment:
I worry that this will no longer do the right thing when we try to differentiate through it. Because the current backwards pass assumes it's always an LCP, it relies on lines in `BoxedLcpConstraintSolver.cpp` like:

```
group.getGradientConstraintMatrices()->registerLCPResults(
        mX,
        hiGradientBackup,
        loGradientBackup,
        fIndexGradientBackup,
        bGradientBackup,
        aColNormGradientBackup,
        aGradientBackup,
        cfm,
        hadToIgnoreFrictionToSolve);
    group.getGradientConstraintMatrices()->constructMatrices(world);
```
Then those constrained groups get used during the computation of `ConstrainedGroupGradientMatrices`'s various Jacobians.

The problem is, if we're using a solver callback, none of that machinery will work.

I think for now, it'd be fine to just flash a big prominent warning if you try to backwards pass, but you've got a `replaceSolveCallback()` in place, like "WARNING: GRADIENTS WILL BE INCORRECT!!!! Nimble is still under heavy development, and we don't yet support differentiating through `timestep()` if you've called `replaceSolveCallback()` to customize the contact model." or something...